flink-formats/flink-protobuf-confluent-registry/src/main/java/org/apache/flink/formats/protobuf/registry/confluent/SchemaCoderProviders.java

```diff
@@ -0,0 +1,337 @@
+        private static final String PACKAGE = "io.confluent.generated";
+        private static final List<Integer> DEFAULT_INDEX = Collections.singletonList(0);
+        /** Subject can be nullable in case coder is only used for deserialization. */
+        private @Nullable final String subject;
+
+        private final ProtobufSchema rowSchema;
+        private final SchemaRegistryClient schemaRegistryClient;
+
+        public DefaultSchemaCoder(
+                @Nullable String subject,
+                RowType rowType,
+                SchemaRegistryClient schemaRegistryClient) {
+            this.subject = subject;
+            rowSchema =
+                    FlinkToProtoSchemaConverter.fromFlinkRowType(
+                            Preconditions.checkNotNull(rowType), ROW, PACKAGE);
+            this.schemaRegistryClient = Preconditions.checkNotNull(schemaRegistryClient);
+        }
+
+        public static MessageIndexes readMessageIndex(DataInputStream input) throws IOException {
+
+            int size = ByteUtils.readVarint(input);
+            if (size == 0) {
+                return new MessageIndexes(DEFAULT_INDEX);
+            } else {
+                List<Integer> indexes = new ArrayList<>(size);
+
+                for (int i = 0; i < size; ++i) {
+                    indexes.add(ByteUtils.readVarint(input));
+                }
+                return new MessageIndexes(indexes);
+            }
+        }
+
+        @Override
+        public ProtobufSchema readSchema(InputStream in) throws IOException {
+            DataInputStream dataInputStream = new DataInputStream(in);
+
+            if (dataInputStream.readByte() != 0) {
+                throw new IOException("Unknown data format. Magic number does not match");
+            } else {
+                int schemaId = dataInputStream.readInt();
+                try {
+                    ProtobufSchema schema =
+                            (ProtobufSchema) schemaRegistryClient.getSchemaById(schemaId);
+                    MessageIndexes indexes = readMessageIndex(dataInputStream);
+                    String name = schema.toMessageName(indexes);
+                    schema = schema.copy(name);
+                    return schema;
+                } catch (RestClientException e) {
+                    throw new IOException(
+                            format("Could not find schema with id %s in registry", schemaId), e);
+                }
+            }
+        }
+
+        @Override
+        public ProtobufSchema writerSchema() {
+            return rowSchema;
+        }
+
+        @Override
+        public void writeSchema(OutputStream out) throws IOException {
+            out.write(CONFLUENT_MAGIC_BYTE);
+            int schemaId = 0;
+            try {
+                schemaId = schemaRegistryClient.register(subject, rowSchema);
+                writeInt(out, schemaId);
+                final ByteBuffer buffer = emptyMessageIndexes();
```

Comment:
We can set empty message indices because the dynamically generated Protobuf Message from Flink RowData will have the root message first. 👍

Reply:
Might be worth adding a comment to clarify that. 

Just note that this is brittle, if the FlinkToProtoSchema code changes such that the message order changes, this will break.

Reply:
Added a comment [here](https://github.com/anupamaggarwal/flink/blob/d0f410636fcf608fecbf530e8dbe6be3ef3e7d02/flink-formats/flink-protobuf-confluent-registry/src/main/java/org/apache/flink/formats/protobuf/registry/confluent/SchemaCoderProviders.java#L147)